File: Archive/src-tauri/src/infrastructure/ffmpeg/render/wave/effects/dot.rs
//! `dot` — vertical dot-matrix columns, fading from centre.
//! Parity: src/waves/dot.ts
use crate::infrastructure::ffmpeg::render::{
    frame::{BAR_FILL, GAP_FILL, WAVE_BARS},
    pixel::fill_circle,
    wave::{WaveCtx, WaveEffect},
};

pub struct DotEffect;

impl WaveEffect for DotEffect {
    fn id(&self) -> &'static str { "dot" }

    fn draw(&self, px: &mut [u8], c: &WaveCtx) {
        let n = WAVE_BARS;
        let dot_rows = 18usize;
        let bar_w = c.ww as f32 * BAR_FILL / n as f32;
        let gap = c.ww as f32 * GAP_FILL / n as f32;
        let row_h = c.wh / dot_rows as f32;
        let dot_r = ((bar_w * 0.38).max(1.5)) as i32;
        for (i, &p) in c.heights.iter().enumerate() {
            let cx = (c.wx as f32 + i as f32 * (bar_w + gap) + bar_w * 0.5) as i32;
            let active = (p * dot_rows as f32 * 0.5).round() as usize;
            for row in 0..dot_rows {
                let cy = (c.wy + (row as f32 + 0.5) * row_h) as i32;
                let dist = (row as i32 - dot_rows as i32 / 2).unsigned_abs() as usize;
                if dist <= active {
                    let intensity = 1.0 - dist as f32 / (dot_rows as f32 / 2.0 + 0.001);
                    fill_circle(px, c.w, c.h, cx, cy, dot_r, c.wc, intensity.max(0.1));
                } else {
                    fill_circle(px, c.w, c.h, cx, cy, (dot_r - 1).max(1), [255, 255, 255], 0.05);
                }
            }
        }
    }
}

File: apps/desktop/src/waves/dot.ts
// `dot` — vertical dot-matrix columns, fading from centre. Parity: effects/dot.rs
import type { WaveEffect } from './types'
import { WAVE_BARS } from './support'
import { BAR_FILL, GAP_FILL } from '../domain/contract.gen'

export const dotEffect: WaveEffect = {
  id: 'dot',
  label: 'LED',
  desc: 'Dot matrix',
  draw({ ctx, color, heights, wx, wy, ww, wh }) {
    const N = WAVE_BARS
    const DOT_ROWS = 18
    const bw = ww * BAR_FILL / N, gap = ww * GAP_FILL / N
    const rowH = wh / DOT_ROWS
    const dotR = Math.max(1.2, Math.min(bw * 0.38, rowH * 0.42))
    heights.forEach((p, i) => {
      const cx = wx + i * (bw + gap) + bw * 0.5
      const active = Math.round(p * DOT_ROWS * 0.5)
      for (let r = 0; r < DOT_ROWS; r++) {
        const cy = wy + (r + 0.5) * rowH
        const dist = Math.abs(r - (DOT_ROWS - 1) / 2)
        if (dist <= active) {
          const intensity = 1 - dist / (DOT_ROWS / 2)
          ctx.fillStyle = color + Math.round(Math.max(0.1, intensity) * 255).toString(16).padStart(2, '0')
          ctx.beginPath(); ctx.arc(cx, cy, dotR, 0, Math.PI * 2); ctx.fill()
        } else {
          ctx.fillStyle = 'rgba(255,255,255,0.05)'
          ctx.beginPath(); ctx.arc(cx, cy, dotR * 0.7, 0, Math.PI * 2); ctx.fill()
        }
      }
    })
  },
}
